hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+                # if block_hash == self.tx_storage.get_from_block_height_index(height):
+                if self.tx_storage.transaction_exists(block_hash):
+                    synced = height
+                    break
+                else:
+                    not_synced = height
+
+        self.log.debug('find_best_common_block finished synced not_synced', synced=synced, not_synced=not_synced)
+        self.synced_height = synced
+
+    def get_peer_block_hashes(self, heights: list[int]) -> Deferred[list[tuple[int, bytes]]]:
+        """ Returns the peer's block hashes in the given heights.
+        """
+        key = 'peer-block-hashes'
+        if self.deferred_by_key.get(key, None) is not None:
+            raise Exception('latest_deferred is not None')
+        self.send_get_peer_block_hashes(heights)
+        deferred: Deferred[list[tuple[int, bytes]]] = Deferred()
+        self.deferred_by_key[key] = deferred
+        return deferred
+
+    def send_get_peer_block_hashes(self, heights: list[int]) -> None:
+        payload = json.dumps(heights)
+        self.send_message(ProtocolMessages.GET_PEER_BLOCK_HASHES, payload)
+
+    def handle_get_peer_block_hashes(self, payload: str) -> None:
+        assert self.tx_storage.indexes is not None
+        heights = json.loads(payload)
+        data = []
+        for h in heights:
+            block = self.tx_storage.indexes.height.get(h)
+            if block is None:
+                break
+            data.append((h, block.hex()))
+        payload = json.dumps(data)
+        self.send_message(ProtocolMessages.PEER_BLOCK_HASHES, payload)
+
+    def handle_peer_block_hashes(self, payload: str) -> None:
+        data = json.loads(payload)
+        data = [(h, bytes.fromhex(block_hash)) for (h, block_hash) in data]
+        key = 'peer-block-hashes'
+        deferred = self.deferred_by_key.pop(key, None)
+        if deferred:
+            deferred.callback(data)
+
+    def send_get_next_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        payload = json.dumps(dict(
+            start_hash=start_hash.hex(),
+            end_hash=end_hash.hex(),
+        ))
+        self.send_message(ProtocolMessages.GET_NEXT_BLOCKS, payload)
+        self.receiving_stream = True
+
+    def handle_get_next_blocks(self, payload: str) -> None:
+        self.log.debug('handle GET-NEXT-BLOCKS')
+        if self._is_streaming:
+            self.protocol.send_error_and_close_connection('GET-NEXT-BLOCKS received before previous one finished')
+            return
+        data = json.loads(payload)
+        self.send_next_blocks(
+            start_hash=bytes.fromhex(data['start_hash']),
+            end_hash=bytes.fromhex(data['end_hash']),
+        )
+
+    def send_next_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        self.log.debug('start GET-NEXT-BLOCKS stream response')
+        # XXX If I don't have this block it will raise TransactionDoesNotExist error. Should I handle this?
```

Comment:
I think so. Maybe just send an empty stream? Or send a "block does not exist" message in the streaming?

Reply:
Done.